### PR TITLE
Size Mapping V2

### PR DESCRIPTION
**Pull request by Fawke**
The core feature has been implemented.
- How to test it?

- Edit the *hello_world.html* file, and replace it with an adUnit object containing the sizeConfig V2 specs. You can get hold of those objects from I2I, here. https://github.com/prebid/Prebid.js/issues/4129

- Currently, the workflow (how to reach and activate this module) is not in place. I've currently created a hook on **getBids** function in adapterManager.js, and that hooked function is getting called for client side adapters only, at the present moment.

- I've disabled the function **checkAdUnitSetup** temporarily in **Prebid.js** file to make this work.

- The **hello_world** example has been modified with a simple adUnit setup with sizeMapping V2 specs to check the code. Later, will create a dedicated html example for this new feature.

- Known Bugs
1. Bidder level sizeConfigs are working weirdly. When sizeConfig disables the bid, the bid is coming as undefined. This should not happen.
2. Haven't yet put checks in place to verify if the structure of the sizeConfig object received is correct or not. So, if we give some object in sizeConfig, not properly configured, the behaviour is unstable.

**EDIT**:

- **Q**: How to test it?

- Check the folder **sizeMappingV2** under *integrationExamples -> gpt*, different scenarios has been already defined there. You can edit them to your liking to test edge cases and other scenarios.

- The Workflow is working End-To-End. The feature is complete.

- **checkAdUnitSetup**  in *prebid.js* is hooked onto in **sizeMappingV2** module to check the *sizeConfig* object.

- All *Known Bugs* mentioned above has been resolved.

- Pending Tasks: 
  - Writing unit tests.
  - Documentation in Prebid.org

#### **Known Issue**

If banner and video mediaTypes, both somehow gets filtered out, **adUnits.sizes** is set to an empty array. Due to this, some bidder adapters may complain.

### CODE REVIEW

**Comment by Fawke** [comment]
@jsnellbaker @jaiminpanchal27 

Modified **checkAdUnitSetup** function to check for the new sizeConfig object.

File: modules/sizeMappingV2.js [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+  let isUsingSizeMappingBool = false;
+  adUnits.forEach(adUnit => {
+    if (adUnit.mediaTypes) {
+      // checks for the presence of sizeConfig property at the adUnit.mediaTypes object
+      Object.keys(adUnit.mediaTypes).forEach(mediaType => {
+        if (adUnit.mediaTypes[mediaType].sizeConfig) {
+          if (isUsingSizeMappingBool === false) {
+            isUsingSizeMappingBool = true;
+            if (auctionId) {
+              _sizeMappingUsageMap[auctionId] = isUsingSizeMappingBool;
```
[the remark this answers is not in the document]
Comment (Fawke) [reply modules/sizeMappingV2.js]: Nice catch! Yes, this function `isUsingNewSizeMapping` will run unnecessarily if it encountered that old size mapping setup is being used. 

Optimally, it should run only once. I've made the required change.

File: modules/sizeMappingV2.js [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+        logError('Detected a mediaTypes.banner object did not include required property sizes. Removing invalid mediaTypes.banner object from request.');
+        delete adUnit.mediaTypes.banner;
+      }
+    } else if (adUnit.sizes) {
+      logWarn('Usage of adUnits.sizes will eventually be deprecated. Please define size dimensions within the corresponding area of the mediaTypes.<object> (eg mediaTypes.banner.sizes).');
+      const bannerSizes = validateSizes(adUnit.sizes);
+      if (bannerSizes.length > 0) {
+        adUnit.sizes = bannerSizes;
+      }
+    }
```
[the remark this answers is not in the document]
Comment (Fawke) [reply modules/sizeMappingV2.js]: Removed it in `sizeMappingV2` file, but may need your help to get rid of it in `prebid.js` file. When I made the required change, around 7 unit test cases failed.

File: modules/sizeMappingV2.js [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+ * @param {MediaTypes} mediaTypes Contains information about supported media types for an Ad Unit and size information for each of those types
+ * @returns {FilteredMediaTypes} Filtered mediaTypes object with relevant media types filtered by size buckets based on activeViewPort size
+ */
+function getFilteredMediaTypes(mediaTypes) {
+  let
+    activeViewportWidth,
+    activeViewportHeight,
+    transformedMediaTypes;
+
+  transformedMediaTypes = Object.assign({}, mediaTypes);
```
[the remark this answers is not in the document]
Comment (Fawke) [reply modules/sizeMappingV2.js]: Made the change.  Thanks for pointing that out.

File: modules/sizeMappingV2.js [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+      // transform mediaTypes object
+      const config = {
+        banner: 'sizes',
+        video: 'playerSize'
+      };
+
+      if (transformedMediaTypes[mediaType].filteredSizeConfig.length > 0) {
+        if (mediaType !== 'native') {
+          transformedMediaTypes[mediaType][config[mediaType]] = transformedMediaTypes[mediaType].filteredSizeConfig[0][config[mediaType]];
+        }
```
[the remark this answers is not in the document]
Comment (Fawke) [reply modules/sizeMappingV2.js]: Added comment.

File: modules/sizeMappingV2.js [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+  return adUnits;
+}
+
+getHook('checkAdUnitSetup').before(function (fn, adUnits) {
+  const usingNewSizeMapping = isUsingNewSizeMapping(adUnits);
+  if (usingNewSizeMapping) {
+    return fn.call(this, adUnits, checkAdUnitSetupHook(adUnits));
+  } else {
+    return fn.call(this, adUnits);
+  }
```
[the remark this answers is not in the document]
Comment (Fawke) [reply modules/sizeMappingV2.js]: Changed it. Now using `fn.bail`

File: modules/sizeMappingV2.js [hunk 6]
```diff
@@ -0,0 +1,406 @@
...
+getHook('getBids').before(function (fn, bidderInfo) {
+  // check if the adUnit is using sizeMappingV1 specs or sizeMappingV2 specs.
+  if (typeof _sizeMappingUsageMap[bidderInfo.auctionId] === 'undefined') {
+    isUsingNewSizeMapping(bidderInfo.adUnits, bidderInfo.auctionId);
+  }
+  if (_sizeMappingUsageMap[bidderInfo.auctionId]) {
+    return fn.call(this, bidderInfo, getBids(bidderInfo));
+  } else {
+    return fn.call(this, bidderInfo);
+  }
```
[the remark this answers is not in the document]
Comment (Fawke) [reply modules/sizeMappingV2.js]: Changed it. Now using fn.bail

File: modules/sizeMappingV2.js [hunk 7]
```diff
@@ -0,0 +1,454 @@
...
+        adUnit = validateBannerMediaType(adUnit);
+      } else if (banner.sizeConfig) {
+        if (Array.isArray(banner.sizeConfig)) {
+          banner.sizeConfig.forEach(config => {
+            // verify if all config objects include "minViewPort" and "sizes" property.
+            // if not, remove the mediaTypes.banner object
+            const keys = Object.keys(config);
+            if (!(includes(keys, 'minViewPort') && includes(keys, 'sizes'))) {
+              logError(`Ad Unit: ${adUnit.code}: mediaTypes.banner.sizeConfig is missing required property minViewPort or sizes or both. Removing the invalid mediaTypes.banner from Ad Unit.`);
+              return delete adUnit.mediaTypes.banner;
```
Comment (jsnellbaker) [inline modules/sizeMappingV2.js]: Given this block of code exists in a `forEach`, if one of the configs were 'bad' - it would still keep looping through even though the parent `mediaTypes.banner` object should have been deleted.  If that's not expected, could you take another look at the logic here (and probably later in the function).

Comment (Fawke) [reply modules/sizeMappingV2.js]: @jsnellbaker Hi, I made the change.
I needed some way to display the log messages, and also delete the banner/video mediaType. I've made a change, let me know if that's fine.

File: modules/sizeMappingV2.js [hunk 8]
```diff
@@ -0,0 +1,464 @@
...
+                    nativeParams: processNativeAdUnitParams(nativeParams)
+                  });
+                }
+
+                bid = Object.assign({}, bid, getDefinedParams(adUnit, ['mediaType', 'renderer']));
+
+                if (bid.sizeConfig) {
+                  const relevantMediaTypes = getRelevantMediaTypesForBidder(bid.sizeConfig, activeViewport);
+                  if (relevantMediaTypes.length === 0) {
+                    logError(`SizeMappingV2:: Bidder: ${bid.bidder} in Ad Unit: ${adUnit.code} has not configured sizeConfig property correctly. This bidder won't be eligible for sizeConfig checks and will remain active.`);
```
Comment (jsnellbaker) [inline modules/sizeMappingV2.js]: This may be a minor thing, but could you flip the order of the adUnit and bidder in these log messages (and below this point)?  

The majority of the messages earlier in the code use adUnit > bidder type of listing, so the switch here to bidder > adUnit is a bit distracting when looking at all the console messages together in the browser.

File: modules/sizeMappingV2.js [hunk 9]
```diff
@@ -0,0 +1,464 @@
...
+                      }, {});
+
+                    if (Object.keys(bidderMediaTypes).length > 0) {
+                      bid = Object.assign({}, bid, { mediaTypes: bidderMediaTypes });
+                    } else {
+                      logInfo(`SizeMappingV2:: Bidder: ${bid.bidder} in Ad Unit: ${adUnit.code} is disabled.`);
+                      return bids;
+                    }
+                  } else {
+                    logInfo(`SizeMappingV2:: Bidder: ${bid.bidder} in Ad Unit: ${adUnit.code} is disabled due to failing sizeConfig check.`);
```
Comment (jsnellbaker) [inline modules/sizeMappingV2.js]: This message should only show when the bidder has the relevantMediaTypes field set to none for this size tier correct?  If so, can we update the message to be more reflecting of that condition?  

The current message feels like they did something wrong in their setup, rather than it be the result of a choice they made.

File: modules/sizeMappingV2.js (hunk 8 above)
Comment (Fawke) [reply modules/sizeMappingV2.js]: Made the change. Please have a look.

File: modules/sizeMappingV2.js (hunk 9 above)
Comment (Fawke) [reply modules/sizeMappingV2.js]: Made the change.
